formats/src/main/scala/org/scanamo/DynamoObject.scala

```diff
@@ -0,0 +1,187 @@
+  }
+
+  final def isEmpty: Boolean = self match {
+    case Empty => true
+    case _     => false
+  }
+
+  final def nonEmpty: Boolean = !isEmpty
+
+  final def contains(key: String): Boolean = self match {
+    case Empty          => false
+    case Strict(xs)     => xs.containsKey(key)
+    case Pure(xs)       => xs.contains(key)
+    case Concat(xs, ys) => xs.contains(key) || ys.contains(key)
+  }
+
+  final def keys: Iterable[String] = self match {
+    case Empty => Iterable.empty[String]
+    case Strict(xs) =>
+      new Iterable[String] {
+        final val iterator = new Iterator[String] {
+          private[this] val underlying = xs.keySet.iterator
+          final def hasNext = underlying.hasNext
+          final def next = underlying.next
+        }
+      }
+    case Pure(xs)       => xs.keys
+    case Concat(xs, ys) => xs.keys ++ ys.keys
+  }
+
+  final def values: Iterable[DynamoValue] = self match {
+    case Empty => Iterable.empty[DynamoValue]
+    case Strict(xs) =>
+      new Iterable[DynamoValue] {
+        final val iterator = new Iterator[DynamoValue] {
+          private[this] val underlying = xs.values.iterator
+          final def hasNext = underlying.hasNext
+          final def next = DynamoValue.fromAttributeValue(underlying.next)
+        }
+      }
+    case Pure(xs)       => xs.values
+    case Concat(xs, ys) => xs.values ++ ys.values
+  }
+
+  final def mapKeys(f: String => String): DynamoObject = self match {
+    case Empty          => Empty
+    case Strict(xs)     => Pure(unsafeToScalaMap(xs).map { case (k, v) => f(k) -> DynamoValue.fromAttributeValue(v) })
+    case Pure(xs)       => Pure(xs.map { case (k, v) => f(k) -> v })
+    case Concat(xs, ys) => Concat(xs.mapKeys(f), ys.mapKeys(f))
+  }
+
+  final def toJavaMap: JMap[String, AttributeValue] = self match {
+    case Empty          => unsafeToJavaMap(Map.empty)
+    case Strict(xs)     => xs
+    case Pure(xs)       => unsafeToJavaMap(xs.mapValues(_.toAttributeValue))
+    case Concat(xs, ys) => unsafeMerge(xs.toJavaMap, ys.toJavaMap)
+  }
+
+  final def toDynamoValue: DynamoValue = DynamoValue.fromDynamoObject(self)
+
+  final def toAttributeValue: AttributeValue =
+    self match {
+      case Empty => DynamoValue.Null
+      case _     => new AttributeValue().withM(toJavaMap)
+    }
+
+  final def toMap[V](implicit D: DynamoFormat[V]): Either[DynamoReadError, Map[String, V]] = self match {
+    case Empty => Right(Map.empty)
+    case Strict(xs) =>
+      xs.entrySet.stream.reduce[Either[DynamoReadError, Map[String, V]]](
+        Right(Map.empty),
+        jfun2[Either[DynamoReadError, Map[String, V]], JMap.Entry[String, AttributeValue], Either[
+          DynamoReadError,
+          Map[String, V]
+        ]]((xs0, e) => for { xs <- xs0; x <- D.read(e.getValue) } yield xs + (e.getKey -> x)),
+        jbiop[Either[DynamoReadError, Map[String, V]]]((xs0, ys0) => for { xs <- xs0; ys <- ys0 } yield xs ++ ys)
```

Comment:
No it isn't. I did that because 2.11 doesn't support SAM types but in fact it does if you add `-Xexperimental` to the compiler flags. Next update will remove all that boilerplate.